loopy/schedule/__init__.py

```diff
@@ -645,6 +654,139 @@ def last_entered_loop(self):
         else:
             return None
 
+# }}}
+
+
+def get_insns_in_topologically_sorted_order(kernel):
+    from pytools.graph import compute_topological_order
+
+    rev_dep_map = {insn.id: set() for insn in kernel.instructions}
+    for insn in kernel.instructions:
+        for dep in insn.depends_on:
+            rev_dep_map[dep].add(insn.id)
+
+    def key(insn_id):
+        # negative of insn.priority because
+        # pytools.graph.compute_topological_order schedules the nodes with
+        # lower 'key' in case of a tie.
+        return (-kernel.id_to_insn[insn_id].priority, insn.id)
+
+    ids = compute_topological_order(rev_dep_map, key=key)
+    return [kernel.id_to_insn[insn_id] for insn_id in ids]
+
+
+# {{{ schedule_as_many_run_insns_as_possible
+
+def schedule_as_many_run_insns_as_possible(sched_state, template_insn):
+    """
+    Returns an instance of :class:`loopy.schedule.SchedulerState`, by appending
+    all reachable instructions that are similar to *template_insn*. We define
+    two instructions to be similar if:
+
+    * Both have the same type.
+    * Both are within the same set of non-parallel inames.
+    * Both belong to the same groups.
+    * Both conflict with the same groups.
+    """
+
+    # {{{ bail when implementation is unsupported
+
+    next_preschedule_item = (
+        sched_state.preschedule[0]
+        if sched_state.preschedule
+        else None)
+
+    if isinstance(next_preschedule_item, (CallKernel, ReturnFromKernel, Barrier)):
```

Comment:
In [!388](https://gitlab.tiker.net/inducer/loopy/-/merge_requests/388) @inducer said:
>What about `EnterLoop` and `LeaveLoop`? Is it OK to continue with those?
Should this be `not isinstance(next_preschedule_item, RunInstruction)` instead?

Reply:
This should be fine because in this case we won't be able schedule any instructions. Logic in the code can be seen [here](https://github.com/inducer/loopy/blob/61c970a7e40474bbf8c561b006a71cd5da6fad50/loopy/schedule/__init__.py#L745-L749).

Adding the extra guard doesn't hurt and would provide a minor performance boost. So I just added it in (9b0a54c).